pyzo/codeeditor/extensions/appearance.py

```diff
@@ -108,14 +109,64 @@ def paintEvent(self, event):
         super(HighlightMatchingOccurrences, self).paintEvent(event)
 
 
+class _ParenNotFound(Exception) :
+    pass
+
+class _ParenIterator :
+    """ Iterates in given direction) over parentheses in the document.
+    Uses the stored token-list of the blocks.
+    Iteration gives both a ParenthesisToken and the global position of
+    the block it lies in (because positions in the ParenthesisToken is
+    relative to the block)."""
+    def __init__(self, cursor, direction) :
+        self.cur_block = cursor.block()
+        self.cur_tokens = self._getParenTokens()
+        self.direction = direction
+        # We need to know where we start in the current token list
+        k = 0
+        try :
+            while self.cur_tokens[k].end != cursor.positionInBlock() :
+                k += 1
+            self.cur_pos = k
+        except IndexError :
+            # If the parenthesis cannot be found, it means that it is not inluded
+            # in any token, ie. it is part of a string or comment
+            raise _ParenNotFound
+    
+    def _getParenTokens(self) :
+        return list(filter(lambda x : isinstance(x, ParenthesisToken), self.cur_block.userData().tokens))
```

Comment:
We cannot really expand the Block class, because its a q Qt thing. I'd feel a bit saver if we wrapped this in a `try: ... except AttributeError: ` block.